sys/dev/ath/if_ath_usb_fw.c

```diff
@@ -0,0 +1,178 @@
+#include <sys/types.h>
+#include <sys/param.h>
+#include <sys/bus.h>
+#include <sys/module.h>
+#include <sys/firmware.h>
+#include <sys/sockio.h>
+#include <sys/kernel.h>
+#include <sys/socket.h>
+#include <sys/endian.h>
+
+#include <net/if.h>
+#include <net/if_dl.h>
+#include <net/if_media.h>
+
+#include <net/if_var.h>
+
+#include <net/if_arp.h>
+#include <netinet/in.h>
+
+#include <net/ethernet.h>
+#include <net80211/ieee80211_freebsd.h>
+#include <net80211/ieee80211_power.h>
+#include <net80211/ieee80211_node.h>
+#include <net80211/ieee80211_var.h>
+#include <net80211/ieee80211_amrr.h>
+#include <net80211/ieee80211_radiotap.h>
+#include <openbsd_adapt.h>
+
+#include "ar5212/ar5212reg.h"
+#include "dev/ath/if_athvar.h"
+
+
+// FreeBSD version
+#include <dev/usb/usb.h>
+#include <dev/usb/usbdi.h>
+
+#include "if_ath_usb_def.h"
+
+#define AR9271_CLOCK_CONTROL		0x050040
+#define AR9271_RESET_POWER_DOWN_CONTROL	0x050044
+#define AR9271_FIRMWARE			0x501000
+#define AR9271_FIRMWARE_TEXT		0x903000
+#define AR7010_FIRMWARE_TEXT		0x906000
+static const struct firmware *fware = NULL;
+
+extern int tsleep_nsec(const void *ident, int priority, const char *wmesg, uint64_t nsecs);
+
+void
+ath_usb_unload_firmware()
+{
+	if (fware == NULL) {
+		printf("Fail: null firmware handler\n");
+		return;
+	}
+
+	firmware_put(fware, FIRMWARE_UNLOAD);
+	printf("Successully called firmware_put\n");
+	return;
+}
+
+int
+ath_usb_get_firmware(struct ath_usb_softc *usc, struct ar_wmi_fw_version *version)
+{
+	usb_device_descriptor_t *dd;
+	const char *name;
+	int error = ENXIO;
+
+	/* Determine which firmware image to load. */
+	if (usc->flags & ATH_USB_FLAG_AR7010) {
+		dd = usbd_get_device_descriptor(usc->sc_udev);
+		name = "ath-open-ar7010.bin";
+	} else
+		name = "athn-open-ar9271.bin";
+	/* Read firmware image from the filesystem. */
+	fware = firmware_get(name);
+	if (fware == NULL) {
+		printf("Failed firmware_get of file %s\n", name);
+		return (error);
+	} else {
+		printf("Success firmware_get of file %s\n", name);
+		printf("Firmware name: %s:\n", fware->name);
+		printf("Firmware version: %u:\n", fware->version);
+		printf("Firmware size: %zu:\n", fware->datasize);
+		version->major = fware->version / 100;
+		version->minor = (fware->version / 10) % 10;
+		return 0;
+	}
+}
+
+int
+ath_usb_transfer_firmware(struct ath_usb_softc *usc)
+{
+	usb_device_request_t req;
+	const char *name;
+	void *ptr;
+	size_t fwsize, size;
+	uint32_t addr;
+	int s, mlen;
+	int error = 0;
+
+	ATH_LOCK(&usc->sc_sc);
+
+/* Load firmware image. */
```

Comment:
No indentation. Cosmetic: can be ignored if no other actions required.

Reply:
Done